File: lib/domain/cultivation_plan/interactors/private_owned_plans_list_interactor.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Interactors
      # 認証ユーザーに紐づく私有計画の一覧（軽量 read）。
      class PrivateOwnedPlansListInteractor
        def initialize(output_port:, user_id:, private_read_gateway:, translator:, logger:, user_lookup:)
          @output_port = output_port
          @user_id = user_id
          @private_read_gateway = private_read_gateway
          @translator = translator
          @logger = logger
          @user_lookup = user_lookup
        end

        def call
          user = @user_lookup.find(@user_id)
          plan_snapshots = @private_read_gateway.list_private_plan_index_plan_snapshots(user_id: user.id)
          plan_ids = plan_snapshots.map(&:id)
          crops_count_hash = @private_read_gateway.count_cultivation_plan_crops_by_plan_ids(plan_ids: plan_ids)
          fields_count_hash = @private_read_gateway.count_cultivation_plan_fields_by_plan_ids(plan_ids: plan_ids)
          plan_row_snapshots = Mappers::PrivatePlanIndexRowsMapper.plan_row_snapshots_with_counts(
            plan_snapshots,
            crops_count_hash: crops_count_hash,
            fields_count_hash: fields_count_hash
          )
          rows = Mappers::PrivatePlanIndexRowsMapper.to_index_rows(plan_row_snapshots)
          @output_port.on_success(rows)
        rescue Domain::Shared::Exceptions::RecordNotFound => e
          @logger.warn("[PrivateOwnedPlansListInteractor] #{e.class}: #{e.message}")
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(@translator.t("plans.errors.session_invalid")))
        rescue NoMethodError, NameError, ArgumentError, SyntaxError
          raise
        rescue Domain::Shared::Exceptions::PersistenceFailed => e
          @logger.error("[PrivateOwnedPlansListInteractor] #{e.class}: #{e.message}")
          raise
        rescue Domain::Shared::Exceptions::RecordInvalid => e
          @logger.warn("[PrivateOwnedPlansListInteractor] #{e.class}: #{e.message}")
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(e.message))
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/interactors/private_owned_plans_list_interactor.rs
//! Ruby: `Domain::CultivationPlan::Interactors::PrivateOwnedPlansListInteractor`

use crate::cultivation_plan::gateways::CultivationPlanPrivateReadGateway;
use crate::cultivation_plan::ports::PrivateOwnedPlansListOutputPort;
use crate::shared::dtos::Error;
use crate::shared::exceptions::{PersistenceFailedError, RecordInvalidError, RecordNotFoundError};
use crate::shared::gateways::UserLookupGateway;
use crate::shared::ports::translator_port::{TranslateOptions, TranslatorPort};
use crate::shared::ports::LoggerPort;

pub struct PrivateOwnedPlansListInteractor<'a, O, G, U, T, L> {
    output_port: &'a mut O,
    user_id: i64,
    private_read_gateway: &'a G,
    translator: &'a T,
    logger: &'a L,
    user_lookup: &'a U,
}

impl<'a, O, G, U, T, L> PrivateOwnedPlansListInteractor<'a, O, G, U, T, L>
where
    O: PrivateOwnedPlansListOutputPort,
    G: CultivationPlanPrivateReadGateway,
    U: UserLookupGateway,
    T: TranslatorPort,
    L: LoggerPort,
{
    pub fn new(
        output_port: &'a mut O,
        user_id: i64,
        private_read_gateway: &'a G,
        translator: &'a T,
        logger: &'a L,
        user_lookup: &'a U,
    ) -> Self {
        Self {
            output_port,
            user_id,
            private_read_gateway,
            translator,
            logger,
            user_lookup,
        }
    }

    pub fn call(&mut self) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let user = self.user_lookup.find(self.user_id);
        match self
            .private_read_gateway
            .list_private_plan_index_rows_by_user_id(user.id)
        {
            Ok(rows) => {
                self.output_port.on_success(rows);
                Ok(())
            }
            Err(err) if err.downcast_ref::<RecordNotFoundError>().is_some() => {
                self.logger.warn(&format!(
                    "[PrivateOwnedPlansListInteractor] RecordNotFound: {err}"
                ));
                let message = self.translator.t(
                    "plans.errors.session_invalid",
                    &TranslateOptions::default(),
                );
                self.output_port.on_failure(Error::new(message));
                Ok(())
            }
            Err(err) if err.downcast_ref::<PersistenceFailedError>().is_some() => {
                self.logger.error(&format!(
                    "[PrivateOwnedPlansListInteractor] PersistenceFailed: {err}"
                ));
                Err(err)
            }
            Err(err) if err.downcast_ref::<RecordInvalidError>().is_some() => {
                let invalid = err.downcast_ref::<RecordInvalidError>().unwrap();
                self.logger.warn(&format!(
                    "[PrivateOwnedPlansListInteractor] RecordInvalid: {err}"
                ));
                let message = invalid
                    .detail_message()
                    .map(|s| s.to_string())
                    .unwrap_or_else(|| invalid.to_string());
                self.output_port.on_failure(Error::new(message));
                Ok(())
            }
            Err(err) => Err(err),
        }
    }
}
